src/main/java/fi/evident/raudikko/WordPart.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * The contents of this file are subject to the Mozilla Public License Version
+ * 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ * https://www.mozilla.org/en-US/MPL/2.0/
+ *
+ * Software distributed under the License is distributed on an "AS IS" basis,
+ * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
+ * for the specific language governing rights and limitations under the
+ * License.
+ *
+ * The Original Code is Libvoikko: Library of natural language processing tools.
+ * The Initial Developer of the Original Code is Harri Pitkänen <hatapitk@iki.fi>.
+ * Portions created by the Initial Developer are Copyright (C) 2012
+ * the Initial Developer. All Rights Reserved.
+ *
+ * Raudikko, the Java port of the Initial Code is Copyright (C) 2020 by
+ * Evident Solutions Oy. All Rights Reserved.
+ *
+ * Alternatively, the contents of this file may be used under the terms of
+ * either the GNU General Public License Version 2 or later (the "GPL"), or
+ * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
+ * in which case the provisions of the GPL or the LGPL are applicable instead
+ * of those above. If you wish to allow use of your version of this file only
+ * under the terms of either the GPL or the LGPL, and not to allow others to
+ * use your version of this file under the terms of the MPL, indicate your
+ * decision by deleting the provisions above and replace them with the notice
+ * and other provisions required by the GPL or the LGPL. If you do not delete
+ * the provisions above, a recipient may use your version of this file under
+ * the terms of any one of the MPL, the GPL or the LGPL.
+ */
+
+package fi.evident.raudikko;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.util.List;
+
+import static fi.evident.raudikko.internal.utils.StringUtils.endsWithChar;
+import static fi.evident.raudikko.internal.utils.StringUtils.startsWithChar;
+
+public interface WordPart {
+
+    @NotNull List<String> getBaseForms();
+    boolean isInBaseForm();
+    boolean isProperNoun();
+
+    @SuppressWarnings("ConstantValue")
+    default @NotNull String toStringWithoutHyphens() {
+        String s = toString();
+        boolean startsWithHyphen = startsWithChar(s, '-');
+        boolean endsWithHyphen = endsWithChar(s, '-');
+
+        if (s.length() == 0 || s.equals("-")) {
+            return "";
+        } else if (!startsWithHyphen && !endsWithHyphen) {
+            return s;
+        } else if (!startsWithHyphen && endsWithHyphen) {
+            return s.substring(0, s.length() - 1);
+        } else if (startsWithHyphen && !endsWithHyphen) {
+            return s.substring(1);
+        } else if (startsWithHyphen && endsWithHyphen) {
+            return s.substring(1, s.length() - 1);
+        } else {
+            throw new IllegalStateException("Error in trimming hyphens");
+        }
+    }
```

Comment:
Tää vaikuttaa vähän epäilyttävältä. Jos ei muuten, niin ainakin nimeämiseltään, koska jotenkin nimen perusteella olettaisi että toteutus olisi ekvivalentti `toString().replace("-", "")` kanssa, mutta ilmeisesti näin ei ole.

Muutenkin, tuonne else-haaraanhan ei ole mahdollista mennä, mutta kääntäjä ei vain tajuta sitä koodin rakenteesta enkä minäkään ensin. Mitä jos toi olisikin:

```java
if (s.isEmpty() || s.equals("-")) 
    return "";

int startOffset = startsWithChar(s, '-') ? 1 : 0;
int endOffset = endsWithChar(s, '-') ? 1 : 0;

return s.substring(startOffset, s.length() - endOffset);
```

Eikös tämä ole semantiikaltaan sama? Ainakin mulle kommunikoi paremmin mistä on kyse. Toi jopa on allokoinneiltaan identtinen, koska String.substring sisältää optimoinnin:

```java
if (beginIndex == 0 && endIndex == length) {
    return this;
}
```

Muutenkin mietin tarvitseeko tuon olla rajapinnassa default-toteutuksena. Ei kai kellään rajapinnan toteuttajalla ole tarvetta overridetä sitä? Harkitsisin ihan vain jotain `removeLeadingAndTrailing(s, '-')` -funktiota. Tuollaisella funktiolle olisi helpompi kirjoittaa testitkin kun ei tule koko muuta roskaa mukana. Ja jos haluaa sen default-toteutuksen sinne rajapintaan, niin sekin voisi olla vain kutsu tuohon staattiseen funktioon.

Reply:
Kyseistä logiikkaa tarvitsee myös pluginin algoritmeissa, joissa käsitellään yhdyssanojen palasia, mutta sen verran pieni se on myös monistaa, joten ehkä se on parempi vain ottaa pois tästä rajapinnasta.